src/main/java/pl/shop/models/Customer.java

```diff
@@ -0,0 +1,46 @@
+package pl.shop.models;
+
+// 'User' is used by spring security, it also feels very generic. Customer fits better =]
+
+
+import lombok.Data;
+import org.springframework.data.annotation.CreatedDate;
+
+import javax.persistence.*;
+import java.time.LocalDate;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+@Entity
+@Data
+public class Customer {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private UUID uuid;
+
+    private String username;
+
+
+    @CreatedDate
+    private LocalDate registeredAt;
+
+    @ManyToMany
+    @JoinTable(
+            name = "customer_products_bought",
+            joinColumns = {@JoinColumn(name = "customer_id")},
+            inverseJoinColumns = {@JoinColumn(name = "product_id")})
+    Set<Product> productsBought = new HashSet<>();
+
+
+    Category favouriteCategory;
+
+//todo
+//    Set<Order> orders = new HashSet<>();
+
+    @OneToOne(cascade = CascadeType.ALL, fetch = FetchType.LAZY, mappedBy = "customer", orphanRemoval = true)
```

Comment:
Niewiele chwilowo rozumiem, ale wygląda kosmicznie, chciałem to tylko podkreślić... Chyba, że da się we w miarę skrótowej wersji wyjaśnić czym są atrybuty w adnotacji, to poproszę. :>

Reply:
hmm, prawdę mówiąc nie wiem czy bardziej czytelne będą wyjaśnienia tutaj czy na discordzie. w razie czego łap mnie tam - jestem przy kompie ~40h/tydzien, zawsze chętnie pogadam. Ale tymczasem, lecimy:

@OneToOne mówi hibernate'owi, że jeden customer może mieć tylko jeden customerDetails i vice versa. Parametry mówią mu o szczegółach tej relacji i tak na prawdę jedynie ułatwiają życie albo poprawiają wydajność, nie są niezbędne do działania bo hibernate sobie zawsze coś tam defaultowo ustawia:

cascade mówi o tym, co ma się stać z drugą stroną relacji, gdy zmienimy pierwszą. Np, jeżeli zmienimy coś w klasie Customer i w jego CustomerDetails, po czym do repo które trzyma Customerów wyślemy metodę .save(customer), to bez kaskady userDetails się nie zapisze (w tej sytuacji musielibyśmy osobno wywołać .save(customerDetails) na repozytorium które trzyma CustomerDetails. Z tym trzeba jednak bardzo ostrożnie, bo jeżeli mamy relację one-to-many albo many-to-many (np product <> order) to możemy w ten sposób skasować sobie przypadkiem całą bazę danych xD (usuwamy order -> on usuwa nam każdy produkt z bazy który się w nim znajdował -> a każdy z tych produktów usuwa każdy order który go zawiera i tak dalej ;) )

fetch są dwie opcje: EAGER i LAZY. Przy EAGER, za każdym razem kiedy wyciągniesz z bazy danych Customera, to razem z nim do pamięci wczytają się jego UserDetails -- nawet jeżeli nic z nimi nie robisz. Przy LAZY, UserDetails wczyta się dopiero kiedy wywołasz na nim jakąś metodę (np getUserDetails() ) [w trakcie trwania sesji hibernate -- czytaj wewnątrz metody opisanej @Transactional]. EAGER fetching prowadzi do problemów wydajnościowych przy większych bazach danych (tu znów to samo co przy cascade -> chcąc wczytać jeden pojedynczy order faktycznie wczytujesz znacznie więcej danych. Bo order sobie fetchuje dane o wszystkich produktach, a każdy z tych produktów fetchuje dane o orderach w których jest)

mappedBy jest akurat prosty - mówi hibernate'owi, że klucz do tej relacji ma się znajdować w tej drugiej encji. Bez tego parametru, Hibernate stworzy foreign key i w Customer (dla customerDetails) i w CustomerDetails (dla Customer)

orphan removal mówi hibernate'owi co ma się stać z drugą stroną relacji, jeżeli z jakiegoś powodu nie będzie miała parenta (np parent zostanie usunięty, albo po prostu ustalimy w Customer pole UserDetails na null). Akurat w tym przypadku ta adnotacja jest zbędna bo i tak mamy CascadeType.ALL)

Reply:
Czyli jak mamy CascadeType.ALL, to usnięcie Customera oznacza usunięcie jego CustomerDetails i w sumie ten orphanRemovel = true jest zbędny właśnie dlatego?

Reply:
Czyli jak mamy CascadeType.ALL, to usnięcie Customera oznacza usunięcie jego CustomerDetails i w sumie ten orphanRemovel = true jest zbędny właśnie dlatego?

Reply:
dokładnie tak. akurat w tym przypadku orphanRemoval ma mało sensu xD Ale, np jeżeli mamy już set którego elementy nie mają sensu jeżeli nie należą do tego zbioru - wtedy już tak (np, koszyk który ma OrderItem (czyli takie dto item+amount)). Element koszyka nie ma sensu jeżeli nie należy do koszyka. A usunąć rzeczy z koszyka mogą różne sytuacje (akcja użytkownika, albo np w trakcie zakupów ktoś inny wykupił ostatnią sztukę z magazynu i wyszło przy ostatecznej walidacji). Dzięki orphanRemoval=true nie musisz się martwić, że wyskoczy Ci jakiś database exception (w przypadku jeśli masz pole orderId not null w OrderItem) / zostaną Ci jakieś bezsensowne osierocone OrderItemy w db (w przypadku przeciwnym). [i przy tej okazji, polecam zwrócić uwagę na przepiękną nazwę tego parametru xD dosłownie - usuwa sieroty xD]

Reply:
Jeszcze dopytam odnośnie mappedBy. Mówisz, że powstrzymuje to tworzenie foreign key dla Customera w CustomerDetails i vice versa. Czy oznacza to, że ten mappedBy jakoś "nadpisuje" konieczność tworzenia relacji między tabelami za pomocą tych kluczy? Ten mappedBy jest tego zamiennikiem?